docs/troubleshoot/app-framework/appserver-error-codes.md

```diff
@@ -2939,303 +2936,353 @@ The following error message codes may appear on the app-server log. Use the foll
 
   **Action:**
 
-  TODO
+  Modify the configuration to use a different keystore type, or migrate the server to z/OS.
 
 
 
-### ZWED0146E":"SAF keyring data had no attribute "%s". Attributes=
+### ZWED0146E
+
+ SAF keyring data had no attribute "_attribute_". Attributes=_attributeKeys_
 
   **Reason:**
 
-  TODO
+  Within the list of _attributeKeys_, _attribute_ could not be found.
 
   **Action:**
 
-  TODO
+  Check the keystore configuration of the server such as in `zowe.certificate` or `components.app-server.node.https` to see if it is valid for Zowe. The SAF keyring Zowe was configured to use may be missing a key and certificate pair, or certificate authorities keychain.
 
 
 
-### ZWED0147E":"SAF keyring data was not found for "%s"
+### ZWED0147E
+
+ SAF keyring data was not found for "_keyName_"
 
   **Reason:**
 
-  TODO
+  The server tried to read the SAF keyring specified within the Zowe configuration, but ran into an error where the server received no data instead.
 
   **Action:**
 
-  TODO
+  Review the logs to see if a reason for the error is shown. Verify that the Zowe configuration points to a valid keyring that the Zowe server user has permissions to read.
 
 
 
-### ZWED0148E":"Exception thrown when reading SAF keyring, e=
+### ZWED0148E
+
+ Exception thrown when reading SAF keyring, e=_error_
 
   **Reason:**
 
-  TODO
+  The SAF keyring which the app-server was configured to use could not be read due to an error. The server likely will not start or will be unable to do any network activity until this error is resolved.
 
   **Action:**
 
-  TODO
+  Review the error message to determine the cause. Often, the error messages will originate from a system service where the documentation can be found here https://www.ibm.com/docs/en/zos/2.5.0?topic=library-return-reason-codes
 
 
 
-### ZWED0149E":"SAF keyring reference missing userId "%s", keyringName "%s", or label "%s"
+### ZWED0149E
+
+ SAF keyring reference missing userId "_user_", keyringName "_name_", or label "_label_"
 
   **Reason:**
 
-  TODO
+  The server configuration specified that the app-server should load keystore and truststore content from a SAF keyring, but the syntax in the configuration was incorrect, because _user_, _name_, or _label_ were not usable by the server.
 
   **Action:**
 
-  TODO
+  Check the `zowe.certificate` or `components.app-server.node.https` sections of Zowe configuration to see if there are entries that start with `safkeyring://` and verify that they are in the format of `safkeyring://USERNAME:RINGNAME&LABEL`. Older versions of zowe will require that there be 4 slashes, such as `safkeyring:////`. The `&LABEL` suffix is only needed for specifying certificate authorities and should be omitted in other sections, for example it is only needed within `zowe.certificate.pem.certificateAuthorities` or `components.app-server.node.https.certificateAuthorities`. 
```

Comment:
Also a good place to link: https://docs.zowe.org/stable/user-guide/configure-certificates-keyring/#create-a-self-signed-jceracfks-certificate